### PR TITLE
SuperPMI: Fix `getFieldType`

### PR DESCRIPTION
**Pull request by CarolEidt**
The `structType` out parameter is optional (i.e. it may be null), but it's not used as a key, so we need to capture it in case other calls with the same keys (`field` and `memberParent`) require it.

### CODE REVIEW

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

**Comment by CarolEidt** [comment]
The other two changes are just incidental cleanup that I found while tracking this down.

**Comment by sandreenko** [comment]
It looks like we have other cases like this, for example, `recGetHelperName` or `recGetAddressOfPInvokeFixup`, so all cases where we have an optional pointer incoming value.

Probably a better general fix would be to add a bool to the key to indicate that. The other fields of the value could depend on `optional != null` in other cases.

How did you find this case?

File: src/ToolBox/superpmi/superpmi-shim-collector/icorjitinfo.cpp [hunk 1]
```diff
@@ -1118,6 +1118,13 @@ CorInfoType interceptor_ICJI::getFieldType(CORINFO_FIELD_HANDLE  field,
                                            )
 {
     mc->cr->AddCall("getFieldType");
+    // 'structType' may be null, but since we use only 'field' and 'memberParent' as keys,
+    // we don't want to pass null, in case we have a subsequent call with non-null 'structType'.
+    CORINFO_CLASS_HANDLE dummyStructType;
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shim-collector/icorjitinfo.cpp]: I don't think this is the correct solution. Fundamentally, we should not alter the function arguments in the collector; we should always pass through what the JIT-passed arguments are intact, and return exactly what the VM returns.

I think in this case perhaps the best option is to change recGetFieldType in the non-null case to first check if the entry (null or not) exists in the LWM (using GetIndex?) and then replacing the stored value.

An alternative that might be better would be to make "structType==nullptr" (the bool condition) a key. Then, we'd store a null and non-null answer.

Comment (AndyAyersMS) [reply src/ToolBox/superpmi/superpmi-shim-collector/icorjitinfo.cpp]: Hmm, when I've implemented these I always have had the shim layer pass all the optional arguments. I suppose if the presence of arguments these alters the logic then that could perturb the results. And that could well be the case.

I wonder if we could have some sort of SPMI replay fidelity check where we diff SPMI replay versus the jit runs that collected the data?

I suppose it comes down to whether the args are altering API behavior or just

**Comment by CarolEidt** [comment]
> How did you find this case?

I did a collection on a pmi of `Microsoft.CodeAnalysis.VisualBasic.dll` and got missing context on the replay.

I'm currently working on the approach of updating the map in the non-null case. I'm not crazy about adding a 3-tuple key.

> It looks like we have other cases like this, for example, recGetHelperName or recGetAddressOfPInvokeFixup, so all cases where we have an optional pointer incoming value.

`getHelperName` doesn't take any optional arguments - `recGetHelperName` is simply handling the case where it returns null, which should be idempotent.

`recGetAddressOfPInvokeFixup` and `recGetPInvokeUnmanagedTarget` are not currently handling optional arguments correctly, but they are described as deprecated in corinfo.h, and are not used by the JIT.

**Pull request merged**